text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,391 @@
+2. Framework compatibility tests are executed.
+3. They can result in one of four states:
+
+    - `pass`: Tests are executed with an expectation to pass. And indeed all of them pass. This means the change is fully compatible with older versions of the CLI.
+    - `pass-on-error`: Tests are executed with expectation to fail with an error message. And indeed the error message is correct.    
+    - `fail`: Tests are executed with an expectation to pass, but some of them don't.
+    - `fail-on-error`: Tests are executed with an expectation to fail with an error message, but the error message is incorrect.
+    
+4. Depending of which state, the developer performs one of the following actions:
+
+    - `make-compatible`: Try to make the code compatible. This would normally happen when the breakage is unintentional, and therefore no version bump was performed.
+    - `mark-incompatible`: Developer reluctantly understands that the changes cannot be compatible, and decides to bump the version. In addition, the test is amended to mark this version as incompatible.
+    - `fix-version-bump`: This would happen when a version bump was performed, but the value it was bumped to is incorrect.
+
+5. If the state of step 3 is `pass` or `pass-on-error`, we are done. Otherwise, *GOTO* 1.
+
+<br>
+
+<!-- To edit: URL Decode -> Edit -> URL Encode -->
+
+<img src='https://g.gravizo.com/svg?digraph%20G%20%7B%0A%20%20%20%20aize%20%3D%224%2C4%22%3B%0A%20%20%20%20%22merge-to-master%22%3B%0A%20%20%20%20%22merge-to-master%22%20-%3E%20%22test-framework-compatibility.sh%22%3B%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22yellowgreen%22%5D%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20pass%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22pass-on-error%22%3B%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22crimson%22%5D%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22fail%22%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22fail-on-error%22%3B%0A%20%20%20%20%22fail%22%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20make-compatible%5CnOR%5Cnmark-incompatible%22%5D%0A%20%20%20%20%22fail-on-error%22%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20fix-version-bump%22%5D%0A%20%20%7D%0A'/>
+
+<br>
+
+# Drawbacks
+
+- Dramatically increases the duration of our integration tests suite.
+- Requires manually editing the framework compatibility test file.
+
+    > This isn't necessarily a drawback, as it provides a clear paper trail of all our breakages.
+
+# Rationale and Alternatives
+
+An alternative approach would to be force CLI versions to *always* be `>=` than Framework Versions. This differs from our current statement where we do try to have newer frameworks support older CLI as much as possible.
+
+Doing so drastically simplify the entire process, as it means we can effectively drop the entire notion of *cx-protocol*.
```

Comment:
Let's do that. I think the main problem we have is actually when framework features depend on new CLI capabilities but developers either fail to bump the protocol version (and there is no fool proof way to enforce it) or the bump is not done correctly because it is tight to the version number of the CDK).

I think we are already at the point where *sometimes* the framework requires that users upgrade the CLI (even if they are not using new features). I would argue that this behavior is even more confusing than just saying: "we expect the CLI to always be >= framework".

By taking this approach we basically solve the two sides of the compatibility challenge:
- CLI is always tested against old framework version(s).
- Framework will never run against an old CLI.

This will eliminate the whole notion of cx protocol version and will not require developers to think about compatibility at all, which is nirvana.

Reply:
I agree, I think the benefits of doing this outweigh the slight, arguable, discomfort it might cause clients. Also, we have already been proven (twice), that users tend to upgrade their CLI first anyway, so I doubt this will have a large affect. 

I am looking to get some more feedback on this though before I incorporate it properly into the RFC.

Reply:
This tears me up. I agree that it simplifies, but I'm not sure how I feel forcing CLI upgrades on users.  It might be fine though, I'm probably being overly conservative.

Reply:
At the moment, we don't treat a CLI upgrade requirement as a big deal. On the contrary, we encourage it by promoting that there is a new version and we have quite a lot of evidence that users upgrade their CLIs eagerly, and ahead of the framework.

Today, every time we introduce a framework feature that requires CLI support, we force (if we remember) a CLI upgrade, even if the user doesn't use the new feature. So users are kind of already accustomed to seeing this message, and we didn't receive much grief about it (we did only when we fucked it up in some other way).

I feel that an assumption that CLI >= framework will dramatically simplify our compatibility story for the product (think CI/CD compat matrix for example), and, together with API compatibility checks and running CLI tests against last framework as proposed in this RFC, will __essentially eliminate all categories of compatibility issues we had__. We have had at least 4 major incidents in the last year related to this.

I think the benefit for users outweighs the "burden" of requiring users to upgrade their CLI. In a sense, since it will now be consistent (i.e. you must upgrade the CLI every time you upgrade the framework, and not just occasionally), we can document it better and users will simply make it a habit.

Side note: In javascript projects, users that use `npx cdk` will also likely have the CLI defined in their project's package.json and will likely be upgraded alongside the framework.

Reply:
I like the simplification of messaging (and code), but I understand @rix0rrr's concern (and share it somewhat).

A side-benefit I see is this would encourage people to consume a CLI locally, which is versioned with the rest of their project. For node it's pretty straight-forward (thanks to `npx` et al) - for other languages it may make sense for us to re-bundle the CLI in idiomatic packages, too (PyPI package, NuGet package, Maven Plug-in, ...) to offer other customers the same mechanisms?

Reply:
Bundling the CLI to idiomatic packages would be brilliant. Though it might not be worth the effort, as we hope to make the CL as slim as possible.